include/alp/base/internalrels.hpp

```diff
@@ -0,0 +1,796 @@
+			 *      all three with length at least \a n. For every i-th element of the
+			 *      three arrays, on the values \f$ x_i, y_i, z_i \f$, \f$ z_i \f$ will
+			 *      be set to \f$ x_i \odot y_i \f$.
+			 *   -# foldrArray, which takes a pointer to an array \f$ x \f$, a
+			 *      parameter \f$ z \f$ of arbitrary type, and a size \n as parameters.
+			 *      The value \f$ z \f$ will be overwritten to \f$ x_i \odot z \f$ for
+			 *      each of the \f$ i \in \{ 0, 1, \ldots, n-1 \} \f$. The order of
+			 *      application, in the sense of which \f$ i \f$ are processed first,
+			 *      is undefined.
+			 *   -# foldlArray, which takes as parameters: \f$ z \f$ of arbitrary type,
+			 *      an array \f$ y \f$, and a size \n. The value \f$ z \f$ will be
+			 *      overwritten to \f$ z \odot y_i \f$ for each of the
+			 *      \f$ i \in \{ 0, 1, \ldots, n-1 \} \f$. The order of application, in
+			 *      the sense of which \f$ i \f$ are processed first, is undefined.
+			 * \endparblock
+			 *
+			 * \note This class only allows wrapping of stateless base operators. This
+			 *       GraphBLAS implementation in principle allows for stateful
+			 *       operators, though they must be provided by a specialised class
+			 *       which directly implements the above public interface.
+			 *
+			 * @see OperatorBase::apply
+			 * @see OperatorFR::foldr
+			 * @see OperatorFL::foldl
+			 * @see \ref OperatorNoFRFLeWiseApply
+			 * @see Operator::foldrArray
+			 * @see Operator::foldlArray
+			 *
+			 * \parblock
+			 * \par Providing New Operators
+			 *
+			 * New operators are easily added to this
+			 * GraphBLAS implementation by providing a base operator and wrapping this
+			 * class around it, as illustrated, e.g., by alp::operators::add as follows:
+			 *
+			 * \snippet ops.hpp Operator Wrapping
+			 *
+			 * This need to be compatible with the GraphBLAS type traits, specifically,
+			 * the #is_operator template. To ensure this, a specialisation of it must be
+			 * privided:
+			 *
+			 * \snippet ops.hpp Operator Type Traits
+			 * \endparblock
+			 */
+			template< typename REL, enum Backend implementation = config::default_backend >
+			class Relation : public RelationBase< REL, implementation > {
+
+				public:
+					typedef typename RelationBase< REL, implementation >::D1 D1;
+					typedef typename RelationBase< REL, implementation >::D2 D2;
+
+			};
+
+			/**
+			 *
+			 * @tparam REL   The generic homogeneous relation.
+			 *
+			 * @see Relation
+			 * @see RelationBase for additional functions exposed to the final relation.
+			 */
+			template< 
+				typename REL, 
+				enum Backend implementation = config::default_backend,
+				std::enable_if_t< 
+					std::is_same< 
+						typename REL::left_type, 
+						typename REL::right_type 
+					>::value 
+				> * = nullptr
+			>
+			class HomogeneousRelation : public Relation< REL, implementation > {
+			};
+
+		} // namespace internal
+
```

Comment:
Are there no type traits for transitive etc.?

I think a single generic type trait would work, meaning a single one for each new algebraic type trait, as follows (psuedocode):

```
template< Relation >
class is_transitive {
    ...value = false;
}

template< Relation, bool = SFINAE< is_relation< Releation > >
class is_transitive {
   ...value = Relation::is_transitive;
};
```

Reply:
yes good point - added.